api/src/wfl/module/covid.clj

```diff
@@ -303,6 +230,137 @@
         (assoc :type terra-workspace-sink-type))
     (throw (ex-info "Invalid sink_items" details))))
 
+
+(defn ^:private add-workload-record
+  "Use `tx` to create a workload `record` for `request` and return the id of the
+   new workload."
+  [tx request]
+  (letfn [(combine-labels [labels]
+            (->> (mapv request [:pipeline :project])
+                 (map str ["pipeline:" "project:"])
+                 (concat labels)
+                 set
+                 sort
+                 vec))]
+    (-> (update request :labels combine-labels)
+        (select-keys [:creator :watchers :labels :project])
+        (merge (select-keys (wfl/get-the-version) [:commit :version]))
+        (assoc :executor "" :output "" :release "" :wdl "" :uuid (UUID/randomUUID))
+        (->> (jdbc/insert! tx :workload) first :id))))
+
+;; Workload Functions
+(defn throw-unless-column-exists [dataset table]
+  "Throw or return the column from `table`"
+  (let [[result & more :as all] (-> table
+                                    (get-in [:columns])
+                                    (->> (filter (comp #{tdr-source-column} :name))))]
+    (when-not result
+      (throw (ex-info "No column with name" {:name tdr-source-column :dataset dataset})))
+    result))
+
+(defn throw-unless-table-exists [dataset]
+  "Throw or return the table from `dataset`"
+  (let [[result & more :as all] (-> dataset
+                                    (get-in [:schema :tables])
+                                    (->> (filter (comp #{tdr-source-table} :name))))]
+    (when-not result
+      (throw (ex-info "No table with name" {:name tdr-source-table :dataset dataset})))
+    (when result
+      (throw-unless-column-exists dataset result))
+    result))
+
+(defn verify-source!
+  "Verify that the `dataset` exists and that the WFL has the necessary permissions to read it"
+  [{:keys [name dataset] :as source}]
+  (when-not (= (:name source) tdr-source-name )
+    (throw (ex-info "Unknown Source" {:source source})))
+  (try
+    (-> (datarepo/dataset (:dataset source))
+        (throw-unless-table-exists))
+    (catch Throwable t
+      (throw (ex-info "Cannot access Dataset" {:dataset dataset
+                                               :cause   (.getMessage t)})))))
```

Comment:
Probably don't want to try/catch here.  It hides the condition raising the original exception.

Reply:
Done